File: backend/app/schemas/sessions.py
"""
sessions.py

Pydantic schemas for WorkoutSession CRUD and nested exercises/sets.
Per DATA_MODEL: reps > 0, weight >= 0, set_number > 0. Exercise order = array index.
"""

from __future__ import annotations

from datetime import date as date_type, datetime
from decimal import Decimal
from typing import List

from pydantic import BaseModel, Field


# --- Set ---


class SetIn(BaseModel):
    """One set in a session exercise (create/update)."""

    set_number: int = Field(..., gt=0, description="Set number within the exercise (1, 2, ...)")
    reps: int = Field(..., gt=0, description="Repetitions in this set")
    weight: Decimal = Field(..., ge=0, description="Weight in lbs")
    rest_seconds: int | None = Field(default=None, ge=0)
    notes: str | None = None


class SetOut(BaseModel):
    """Set as returned in API responses."""

    id: int
    set_number: int
    reps: int
    weight: Decimal
    rest_seconds: int | None = None
    notes: str | None = None

    model_config = {"from_attributes": True}


# --- Workout exercise (nested in session) ---


class WorkoutExerciseIn(BaseModel):
    """Exercise + sets in a session (create/update). Order = array index."""

    exercise_id: int = Field(..., gt=0)
    sets: List[SetIn] = Field(default_factory=list)
    notes: str | None = None


class WorkoutExerciseOut(BaseModel):
    """Exercise with sets as returned in session responses."""

    exercise_id: int
    exercise_name: str
    sets: List[SetOut] = Field(default_factory=list)
    notes: str | None = None

    model_config = {"from_attributes": True}


# --- Session ---


class SessionCreate(BaseModel):
    """Request body for POST /sessions."""

    date: date_type
    notes: str | None = None
    exercises: List[WorkoutExerciseIn] = Field(default_factory=list)


class SessionUpdate(BaseModel):
    """Request body for PUT /sessions/{session_id}."""

    date: date_type | None = None
    notes: str | None = None
    exercises: List[WorkoutExerciseIn] | None = None


class SessionOut(BaseModel):
    """Session as returned in get/list (with nested exercises and sets)."""

    id: int
    date: date_type
    notes: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    exercises: List[WorkoutExerciseOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class SessionListResponse(BaseModel):
    """Response for GET /sessions: sessions and total count."""

    sessions: List[SessionOut] = Field(default_factory=list)
    total: int = 0


class LastSetPoint(BaseModel):
    """One set from a previous session (for pre-fill)."""

    set_number: int
    weight: Decimal
    reps: int


class LastSetsOut(BaseModel):
    """Last time this exercise was logged: date and sets (for pre-fill)."""

    date: date_type
    sets: List[LastSetPoint] = Field(default_factory=list)

File: frontend/src/api/sessions.ts
/**
 * Sessions API: list, create, get, update, delete.
 * Handles 409 (duplicate date) via error response; caller should show detail message.
 */
import { apiClient } from "./client";

const SESSIONS_PREFIX = "/api/v1/sessions";

export interface SetOut {
  id: number;
  set_number: number;
  reps: number;
  weight: number;
  rest_seconds?: number | null;
  notes?: string | null;
}

export interface WorkoutExerciseOut {
  exercise_id: number;
  exercise_name: string;
  sets: SetOut[];
  notes?: string | null;
}

export interface SessionOut {
  id: number;
  date: string;
  notes: string | null;
  created_at?: string;
  updated_at?: string;
  exercises: WorkoutExerciseOut[];
}

export interface SessionListResponse {
  sessions: SessionOut[];
  total: number;
}

export interface SetIn {
  set_number: number;
  reps: number;
  weight: number;
  rest_seconds?: number | null;
  notes?: string | null;
}

export interface WorkoutExerciseIn {
  exercise_id: number;
  sets: SetIn[];
  notes?: string | null;
}

export interface SessionCreateInput {
  date: string;
  notes?: string | null;
  exercises: WorkoutExerciseIn[];
}

export interface SessionUpdateInput {
  date?: string;
  notes?: string | null;
  exercises?: WorkoutExerciseIn[];
}

export interface ListSessionsParams {
  limit?: number;
  offset?: number;
  start_date?: string;
  end_date?: string;
}

export async function listSessions(params?: ListSessionsParams): Promise<SessionListResponse> {
  const { data } = await apiClient.get<SessionListResponse>(`${SESSIONS_PREFIX}/`, {
    params: params ?? {},
  });
  return data;
}

export async function getSession(id: number): Promise<SessionOut> {
  const { data } = await apiClient.get<SessionOut>(`${SESSIONS_PREFIX}/${id}`);
  return data;
}

export async function createSession(input: SessionCreateInput): Promise<SessionOut> {
  const { data } = await apiClient.post<SessionOut>(`${SESSIONS_PREFIX}/`, input);
  return data;
}

export async function updateSession(id: number, input: SessionUpdateInput): Promise<SessionOut> {
  const { data } = await apiClient.put<SessionOut>(`${SESSIONS_PREFIX}/${id}`, input);
  return data;
}

export async function deleteSession(id: number): Promise<void> {
  await apiClient.delete(`${SESSIONS_PREFIX}/${id}`);
}

/** Last time this exercise was logged: date and sets (for pre-fill). 404 if never logged. */
export interface LastSetPoint {
  set_number: number;
  weight: number;
  reps: number;
}

export interface LastSetsOut {
  date: string;
  sets: LastSetPoint[];
}

export async function getLastSetsForExercise(exerciseId: number): Promise<LastSetsOut | null> {
  try {
    const { data } = await apiClient.get<LastSetsOut>(`${SESSIONS_PREFIX}/last-sets`, {
      params: { exercise_id: exerciseId },
    });
    return data;
  } catch {
    return null;
  }
}

/** Format date as YYYY-MM-DD for API. */
export function formatDateForApi(d: Date): string {
  const year = d.getFullYear();
  const month = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

/** Today in YYYY-MM-DD. */
export function todayApi(): string {
  return formatDateForApi(new Date());
}
